library/entity/interaction/src/main/java/org/quiltmc/qsl/entity/interaction/mixin/LivingEntityMixin.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.entity.interaction.mixin;
+
+import net.minecraft.entity.Entity;
+import net.minecraft.entity.EntityType;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.entity.damage.DamageSource;
+import net.minecraft.world.World;
+import org.quiltmc.qsl.entity.interaction.api.LivingEntityAttackCallback;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+@Mixin(LivingEntity.class)
+public abstract class LivingEntityMixin extends Entity {
+
+	@Inject(method = "damage", at = @At("HEAD"), cancellable = true)
+	private void onTakeDamage(DamageSource source, float amount, CallbackInfoReturnable<Boolean> cir) {
+		if (source.getAttacker() instanceof LivingEntity attacker) {
+			boolean result = LivingEntityAttackCallback.EVENT.invoker().onAttack(attacker, this, source, amount);
```

Comment:
What if someone wanted to modify the amount of damage dealt to the entity?

Reply:
I've been trying to think of a way to best do this while still being able to cancel the method and possibly pass the new damage value from one listener to the next.

Reply:
A shared context could work. Like a mutable object that you can pass in as a callback parameter that's just gonna contain 1. damage amount and 2. whether the callback was cancelled or not.
This way all listeners would get the modified damage from *all* the previous ones and anyone can cancel.

Reply:
I can make a draft if you want.

Reply:
Something like this
![image](https://user-images.githubusercontent.com/46004440/186456429-a067c2fa-2794-45d8-bea4-0e2f13c0d053.png)

Reply:
If you would I'd appreciate it. I'm curious how you would set the damage in the method while still still remaining cancellable. I've considered `@ModifyVariable` or `@ModifyArg` but I don't believe you can cancel a method with them. I was going to look into possibly multiple injects, but I'm not sure.

Reply:
> Something like this ![image](https://user-images.githubusercontent.com/46004440/186456429-a067c2fa-2794-45d8-bea4-0e2f13c0d053.png)

Thx, the context thing itself isn't what I'm struggling with. I'm struggling with actually setting the new damage value in the mixin while also remaining cancellable. I think I have a working system that is kinda scuffed imo. I'm going to get some feedback from the toolchain discord before I push it to my repo.

Thanks for the suggestions!

Reply:
I am worried about the way this would be handled.
The issue with the cancellability is how do you listen for something and act upon it without it beign canceled? since if it cancels your action could still trigger and that's an issue.

Reply:
Two-phase event, one for possible cancellations, second to act in case event has not been cancelled?

Reply:
Alternatively, one pass, but collect a list of deferred actions and final cancellation results they should be executed on from every listener